tests/cupy_tests/core_tests/test_ndarray.py

```diff
@@ -39,44 +40,77 @@ def wrap_take(array, *args, **kwargs):
     return array.take(*args, **kwargs)
 
 
-@testing.parameterize(
-    {'arg': None, 'shape': ()},
-    {'arg': 3, 'shape': (3,)},
-)
 @testing.gpu
 class TestNdarrayInit(unittest.TestCase):
 
-    def test_shape(self):
-        a = core.ndarray(self.arg)
-        self.assertTupleEqual(a.shape, self.shape)
+    def test_shape_none(self):
+        a = cupy.ndarray(None)
+        assert a.shape == ()
 
+    def test_shape_int(self):
+        a = cupy.ndarray(3)
+        assert a.shape == (3, )
 
-@testing.gpu
-class TestNdarrayOrder(unittest.TestCase):
+    def test_memptr(self):
+        a = cupy.arange(6).astype(numpy.float32).reshape((2, 3))
+        memptr = a.data
+
+        b = cupy.ndarray((2, 3), numpy.float32, memptr)
+        testing.assert_array_equal(a, b)
+
+        b += 1
+        testing.assert_array_equal(a, b)
+
+    def test_memptr_with_strides(self):
+        a = cupy.arange(12).astype(numpy.float32).reshape((2, 6))[::-1, ::2]
+        memptr = a.data
+
+        b = cupy.ndarray((2, 3), numpy.float32, memptr, strides=a.strides)
+        testing.assert_array_equal(a, b)
+        assert a.strides == b.strides
+
+        b += 1
+        testing.assert_array_equal(a, b)
+
+    def test_strides(self):
+        a = cupy.ndarray((2, 3), numpy.float32, strides=(8, 4))
```

Comment:
How about giving `cupy.ndarray(20, numpy.uint8)` to `memptr`, to avoid the potential issue on the memory allocation.  (See my comment on `core.pyx`.)

Reply:
It would be duplicated with `test_memptr_with_strides`.
Can we just replace this test with a check to ensure the error is raised?

Reply:
I would just rename the test to `test_memptr_with_strides_2`, because this `a` cannot be created by other means.

I agree with that we need a test to assert the error is raised.  Could you add a new test?
For example `@numpy_cupy_raises()` on `xp.ndarray((2, 3), numpy.float32, strides=(20, 4))`

Reply:
Since the method `_set_shape_and_strides` was private, this is the first time to support self-overwrapping arrays that are not merely broadcasted.  I'd like to test such arrays.

On the creation, `numpy` raises if the buffer is too small:
```
>>> np.ndarray((2, 3), 'f', np.arange(5, dtype='f').data)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: buffer is too small for requested array
>>> np.ndarray((2, 3), 'f', np.arange(5, dtype='f').data, strides=(12, 4))  # C-contiguous strides
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: strides is incompatible with shape of requested array and size of buffer
>>> np.ndarray((2, 3), 'f', np.arange(5, dtype='f').data, strides=(8, 4))
array([[0., 1., 2.],
       [2., 3., 4.]], dtype=float32)
```
I think it's OK to leave the size of `memptr` unchecked, while there should not be a wrong check.

Reply:
> OK, then, how about replacing the configuration of test_memptr_with_strides with overlapping one?

:+1:

Reply:
Fixed.